Migrate RCBackend

taquitos opened the pull request:
Fixes #581

Juanpe commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,614 @@
+        }
+    }
+
+    private func handleLogin(response: [String: Any]?,
```
since there are so many `private` methods, they could be extracted to a `private extension`. WDYT?

taquitos replied:
Love it.

Juanpe commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,614 @@
+                let nonce: UUID?
+                if let nonceString = nonceString {
+                    nonce = UUID(uuidString: nonceString)
+                } else {
```
[totally minor] Is this `else` block needed?

taquitos replied:
Unfortunately, yeah, the `let nonce: UUID?` on line 200 isn't set yet, so it won't compile we don't explicitly set it.

Juanpe replied:
oohh, you're right,  If it was a `var`, it would be possible. I didn't realize that `nonce` is a constant 🤦🏻‍♂️

Juanpe commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,614 @@
+                                            ]
+                                          ]]
+
+        self.httpClient.performPOSTRequest(serially: true, path: "/offers",
```
[Style] `path` should be in another line, shouldn't be?

taquitos replied:
Oops

Juanpe commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,614 @@
+                                      path: path,
+                                      requestBody: body,
+                                      headers: authHeaders) { statusCode, response, error in
+            self.handle(response: response, statusCode: statusCode, error: error, completion: completion)
```
[Style] the parameters should be one per line, shouldn't be?

taquitos replied:
Ooops

Juanpe commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,614 @@
+                                      requestBody: requestBody,
+                                      headers: authHeaders) { statusCode, response, error in
+
+            self.handleLogin(response: response, statusCode: statusCode, error: error, completion: completion)
```
[Style] the parameters should be one per line, shouldn't be?

Juanpe commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,614 @@
+            let maybeCallbacks = purchaserInfoCallbacksCache[key]
+            var callbacks: [BackendPurchaserInfoResponseHandler]
+            let requestAlreadyInFlight: Bool
+            if maybeCallbacks == nil {
+                requestAlreadyInFlight = false
+                callbacks = []
+            } else {
+                requestAlreadyInFlight = true
+                callbacks = maybeCallbacks!
+            }
```
[Just a suggestion] In order to avoid the force unwrapping (maybe in the future, the rule `force_unwrapping` could be activated), we could use `if let` instead. WDYT?
```suggestion
            if let someCallbacks = maybeCallbacks {
                requestAlreadyInFlight = true
                callbacks = callbacks
            } else {
                requestAlreadyInFlight = false
                callbacks = []
            }
```

taquitos replied:
I agree, in general `force_unwrapping` is something to be avoided. However, if you can prove that for unwrapping won't cause problems, it's ok. That being said, in this case, your suggestion removes the need for force unwrapping and is better, so I'll go with that!

Juanpe replied:
totally agree. Feel free to leave this code there 🙂

Juanpe commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,614 @@
+    private let apiKey: String
+
+    // callbackQueue controls access to both offeringsCallbacksCache and purchaserInfoCallbacksCache
+    private let callbackQueue = DispatchQueue(label: "Backend callbackQueue")
```
Would it make sense to have one queue per dictionary? Actually, the code is correct but I'm thinking about a scenario where this could fail 🤔

taquitos replied:
I think in general, yes. In this case, we already spin up a few queues for locking purposes. Since it's not necessary to have two (one won't impact performance/correctness), and one queue reduces complexity, I think keeping it to one is best.

Juanpe commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,614 @@
+            let message = response?["message"] as? String
+            let responseError = ErrorUtils.backendError(withBackendCode: code, backendMessage: message)
+            completion?(responseError)
+            return
```
this `return` is useless, doesn't it?

taquitos replied:
So, yes, currently, it's useless 😄  I put it in because there is always a risk that the next engineer could add more code below this so it makes people (probably me) think a little more.
That being said, I'm going to refactor this a tiny bit and use a `guard` because I think that will convey the intent better.

Juanpe commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,614 @@
+    }
+
+    private func escapedAppUserID(appUserID: String) -> String {
+        return appUserID.addingPercentEncoding(withAllowedCharacters: NSMutableCharacterSet.urlHostAllowed)!
```
[Style] omit `return`?

taquitos replied:
Refactoring to use `guard`

Juanpe commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,614 @@
+    private func getOfferingsCallbacksAndClearCache(forKey key: String) -> [OfferingsResponseHandler] {
+        return callbackQueue.sync { [self] in
+            let callbacks = offeringsCallbacksCache.removeValue(forKey: key)
+            // TODO: Should we throw instead of NSParameterAssert?
```
👌🏼

taquitos replied:
😄

Juanpe reviewed (commented):
Hi @taquitos, I added some small comments. Most of them are related to the swift style guide 🙂

Juanpe commented:
For sure it's out of the scope of this ticket, but I was thinking about this huge file and I think we could extract some logic. This class is like a facade of the backend including all the available operations. So, an approach could be to create objects that encapsulate each backend request/operation, they could be called`BackendOperation`, `BackendResource`, `NetworkRequest`... Then we could have operations like `LogInOperation`, `GetOfferingsOperation`, etc.

WDYT? Would it make sense?

taquitos commented:
re: `create objects that encapsulate each backend request/operation` yes, 💯  yes 😄  This is exactly what I would do. It would reduce the parameter count, make changing the api easier, make calling the api easier, and enable people to collaborate easier across the api boundaries. When I was interviewing at RC, I actually gave folks a presentation on the subject of doing just this in a previous project 😆

Juanpe commented:
> re: `create objects that encapsulate each backend request/operation` yes, 💯 yes 😄 This is exactly what I would do. It would reduce the parameter count, make changing the api easier, make calling the api easier, and enable people to collaborate easier across the api boundaries. When I was interviewing at RC, I actually gave folks a presentation on the subject of doing just this in a previous project 😆

Awesome!! 🎉 I used this approach for handling the deep-links on my current project and the team love it 😄 So...where I could apply?😜 😂

Juanpe commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -541,16 +545,29 @@ public typealias OfferSigningResponseHandler = (String?, String?, UUID?, NSNumbe
             if let someCallbacks = maybeCallbacks {
                 requestAlreadyInFlight = true
-                callbacks = callbacks
+                callbacks = someCallbacks
```
🤦🏻‍♂️ 🙏🏼

taquitos replied:
lol

Juanpe commented:
Thanks for the changes 🤙

taquitos commented:
re: `So...where I could apply?😜 😂`
[https://boards.greenhouse.io/revenuecat](https://boards.greenhouse.io/revenuecat)

![EyeBrows](https://user-images.githubusercontent.com/1304821/128400948-6b8359ed-f9d7-4c58-81fd-6eb446126f2f.gif)

taquitos commented:
Also added: https://github.com/RevenueCat/purchases-ios/issues/715
as a follow up. I addressed a couple places where we weren't input checking on the `appUserID`, and one place where we should be returning an error if the `appUserID` is invalid-> https://github.com/RevenueCat/purchases-ios/pull/713/files#diff-2f16e871301b568ccf8407b2ab43e2faa1fec56fdc1fe4260d908f8849f8c1cfR362

aboedo commented on PurchasesCoreSwift/Attribution/AttributionData.swift:
```diff
@@ -11,11 +11,11 @@ import Foundation
 @objc(RCAttributionData) public class AttributionData: NSObject {
 
-    @objc public let data: [AnyHashable: Any]
+    @objc public let data: [String: Any]
```
🙌

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+
+    @objc(createAliasForAppUserID:newAppUserID:completion:)
+    public func createAlias(appUserID: String, newAppUserID: String, completion: ((Error?) -> Void)?) {
+        guard let appUserID = try? escapedAppUserID(appUserID: appUserID) else {
+            completion?(ErrorUtils.missingAppUserIDError())
+            return
```
👍

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+        httpClient.clearCaches()
+    }
+
+    // swiftlint:disable function_parameter_count
```
[not for this PR] we should probably just start creating structs for these when they get too big. `ProductInfo` was actually meant to help, but there's still a bunch more params

taquitos replied:
💯

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+                     completion: @escaping BackendPurchaserInfoResponseHandler) {
+    // swiftlint:enable function_parameter_count
+        let fetchToken = receiptData.base64EncodedString(options: .lineLength64Characters)
+        var body: [String: AnyObject] = [
```
I think the conversion into `NSObject`s isn't needed now that `HTTPClient` is implemented in Swift

aboedo replied:
we can also go ahead and make the `HTTPClient` `internal` and remove the `@objc` modifiers from its methods

taquitos replied:
Will do it in another PR

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+        -\(offeringIdentifier ?? "")-\(observerMode)-\(subscriberAttributesByKey?.debugDescription ?? "")"
+        """
+
+        if add(callback: completion, key: cacheKey) {
```
I realize that this is an exact port of what we had before, but I had to go in and check the implementation to understand what it does. 

Perhaps we could rename this method? I'm honestly struggling to come up with a good name too, since this adds the value to the completion blocks for the key if needed and returns a bool stating whether a request was already in flight. 

`addCallbackToListAndReturnIfRequestNotInFlight` doesn't quite roll off the tongue or describe what's happening. 

what if we had a private enum, i.e.: 
```swift
let result = appendToCallbacksList(callback, forKey: cacheKey)
if result == .requestAlreadyInFlight { 
    return
}
```

taquitos replied:
I'll split the difference.

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+            body.merge(productInfo.asDictionary()) { _, new in new }
+        }
+
+        if let subscriberAttributesByKey = subscriberAttributesByKey {
+
+            let attributesInBackendFormat = subscriberAttributesToDict(subscriberAttributes: subscriberAttributesByKey)
```
FFTI: is the extra line break intentional?

taquitos replied:
Oops

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+
+        httpClient.performGETRequest(serially: true, path: path, headers: authHeaders) { statusCode, response, error in
+            for completion in self.getPurchaserInfoCallbacksAndClearCache(forKey: path) {
+                print("Hi")
```
👋

taquitos replied:
〰️

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+        httpClient.performGETRequest(serially: true, path: path, headers: authHeaders) { statusCode, response, error in
+            for completion in self.getPurchaserInfoCallbacksAndClearCache(forKey: path) {
+                print("Hi")
+                self.handle(purchaserInfoResponse: response,
+                            statusCode: statusCode,
+                            error: error,
+                            completion: completion)
```
note, just for the sake of discussion: this might be a case for a `weak self`:

If we're getting subscriber info, but the `Backend` instance has been dealloc'ed, it's likely that the SDK has been re-configured, and it might have a different user. 
In a case like this, we might not want to overwrite the data in cache with a new value from a potentially old user. 

However, I think deciding that shouldn't be up to the `Backend` class, but rather the `IdentityManager` or even `PurchaserInfoManager`. So I think the strong self here is fine.

taquitos replied:
This is a great catch. I'm going back through the this class and seeing where it makes sense for `weak self`. I actually had a single failing test related to re-configuring the SDK, and switching it to `weak self` fixes it.

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+                     appUserID: String,
+                     completion: @escaping OfferSigningResponseHandler) {
+    // swiftlint:enable function_parameter_count
+        let fetchToken = receiptData.base64EncodedString(options: .lineLength64Characters)
```
perhaps we could extract this logic since we have it in a couple of places?

taquitos replied:
💯

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+                let signature = signatureData["signature"] as? String
+                let keyIdentifier = offer["key_id"] as? String
+                let nonceString = signatureData["nonce"] as? String
+                let nonce: UUID?
```
`maybeNonce`

taquitos replied:
✅

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+                     appUserID: String,
+                     completion: ((Error?) -> Void)?) {
+        guard subscriberAttributes.count > 0 else {
+            Logger.warn(Strings.attribution.empty_subscriber_attributes)
+            return
```
we should probably still call completion here. I think this was a bug in the previous implementation

taquitos replied:
Yeah, the thing I was thinking about this piece was, is this an error condition? It doesn't seem like a hard error, but more like "hey, you might have expected attributes but there are none". That being said, calling the completion with a nil error also seems wrong.

taquitos replied:
I decided to create another ErrorCode case

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+            return
+        }
+
+        if receiptData.count == 0 {
```
nitpick: `receiptData.isEmpty()`

aboedo replied:
🤔 kinda feels like this bit is out of scope for the Backend class, so whenever we do https://github.com/RevenueCat/purchases-ios/issues/690 we should probably move this over. 
I'll add a note there

taquitos replied:
Yeah, thanks!

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+            return
+        }
+
+        // Closure we can use for both missing appUserID as well as server error where we have an unknown
+        // eligibility status.
+        let unknownEligibilityClosure: () -> [String: IntroEligibility] = {
```
perhaps we can have a `handleResult` method like we do for some of the others?

taquitos replied:
💯

aboedo reviewed (commented):
great job on this! It's hard to convey how much nicer this is to read ❤️ 

I'm still reviewing, but I'm wrapping up for the day so I'm leaving some partial comments

aboedo commented on PurchasesTests/Networking/BackendTests.swift:
```diff
@@ -169,7 +179,7 @@ class BackendTests: XCTestCase {
         let isRestore = arc4random_uniform(2) == 0
         let observerMode = arc4random_uniform(2) == 0
 
-        backend?.postReceiptData(receiptData,
+        backend?.post(receiptData: receiptData,
                                  appUserID: userID,
                                  isRestore: isRestore,
                                  productInfo: nil,
```
indentation for these is messed up now

taquitos replied:
SO MANY MESSED UP 🤦

aboedo commented on PurchasesTests/Networking/BackendTests.swift:
```diff
@@ -718,6 +731,46 @@ class BackendTests: XCTestCase {
         expect(eligibility!["productc"]!.status).toEventually(equal(IntroEligibilityStatus.unknown))
     }
 
+    func testEligibilityUnknownIfMissingAppUserID() {
```
👏👏👏👏

taquitos replied:
👯

aboedo commented on PurchasesTests/Networking/BackendTests.swift:
```diff
@@ -718,6 +731,46 @@ class BackendTests: XCTestCase {
+        var underlyingError = (eventualError?.userInfo[NSUnderlyingErrorKey]) as? NSError
+        expect(underlyingError).to(beNil())
+
+        backend?.getIntroEligibility(appUserID: "   ", receiptData: Data.init(1...2), productIdentifiers: products, completion: {(productEligibility, error) in
```
let's split this one up into one line per param

taquitos replied:
There were a few like this I took care of.

aboedo commented on PurchasesTests/Networking/BackendTests.swift:
```diff
@@ -718,6 +731,46 @@ class BackendTests: XCTestCase {
+
+        // We only have an underlyingError if we got a response from the server. This shouldn't happen because
+        // we shouldn't even send a request.
+        var underlyingError = (eventualError?.userInfo[NSUnderlyingErrorKey]) as? NSError
```
this can be `let`, right?

aboedo replied:
I usually recommend not commenting code, but coding comments instead, i.e.:

```swift
let errorComingFromBackend = (eventualError?.userInfo[NSUnderlyingErrorKey]) as? NSError
let wasRequestSent = errorComingFromBackend != nil
expect(wasRequestSent) == false
```

taquitos replied:
This is a good point, it helps keep documentation up-to-date if it's built into the code.

Juanpe commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,669 @@
+            }
+            callbacks.append(callback)
+            purchaserInfoCallbacksCache[key] = callbacks
+            return requestAlreadyInFlight
```
I agree 😅  Personally I really like the second one 👌🏼

taquitos replied:
✅  I also added an enum instead of returning a bool, I think the calling logic makes more sense.

taquitos commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+
+}
+
+private extension Backend {
```
Yeah, this was a great suggestion by @juanpe

taquitos commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+            return
+        }
+
+        let created = statusCode == 201
```
Just added `.createdSuccess` to the enum

taquitos commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+        let successfullySynced = !(isInternalServerError || isNotFoundError)
+        resultDict[Backend.RCSuccessfullySyncedKey as String] = NSNumber(value: successfullySynced)
+
+        let hasAttributesResponseContainerKey = (response[Backend.RCAttributeErrorsResponseKey] != nil)
+        let attributesResponseDict = hasAttributesResponseContainerKey
+            ? response[Backend.RCAttributeErrorsResponseKey]
+            : response
```
In spirit 😄 
`attributesResponse = (response[Backend.RCAttributeErrorsResponseKey] as? [String: AnyObject]) ?? response`

taquitos commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+            ? response[Backend.RCAttributeErrorsResponseKey]
+            : response
+
+        if let attributesResponseDict = attributesResponseDict as? [String: Any] {
+            let hasAttributeErrors = (attributesResponseDict[Backend.RCAttributeErrorsKey] != nil)
+            if hasAttributeErrors {
+                resultDict[Backend.RCAttributeErrorsKey] = attributesResponseDict[Backend.RCAttributeErrorsKey] as AnyObject
+            }
+        }
```
This does look nicer ✅

taquitos commented on PurchasesCoreSwift/Networking/HTTPRequest.swift:
```diff
@@ -36,7 +36,7 @@ public class HTTPRequest: NSObject, NSCopying {
                          retried: Bool,
                          urlRequest: URLRequest,
-                         completionHandler: ((Int, [AnyHashable: Any]?, Error?) -> Void)?) {
+                         completionHandler: ((Int, [String: Any]?, Error?) -> Void)?) {
```
😻

taquitos commented:
Rebased/merged

taquitos commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,650 @@
+    }
+
+    func attributesUserInfoFromResponse(response: [String: Any], statusCode: Int) -> [String: AnyObject] {
+        var resultDict: [String: AnyObject] = [:]
```
Updated!

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,664 @@
+        -\(offeringIdentifier ?? "")-\(observerMode)-\(subscriberAttributesByKey?.debugDescription ?? "")"
+        """
+
+        if add(callback: completion, key: cacheKey) == .addedToExistingInFlightList {
+            return
```
👍

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,664 @@
+        for productID in response.productIdentifiers {
+            let status: IntroEligibilityStatus
+
+            if let e = responseData[productID] as? Bool {
+                status = e ? .eligible : .ineligible
```
I'd still name this `eligibility` because I'm not a huge fan of single-letter variables

taquitos replied:
✅

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,664 @@
+            responseData = [:]
+        }
+
+        guard let responseData = responseData else {
```
we're assuming a very specific format for the `responseData`, so I'd name it accordingly as `eligibilitiesByProductIdentifier`

taquitos replied:
👍

aboedo commented on PurchasesCoreSwift/Networking/Backend.swift:
```diff
@@ -0,0 +1,664 @@
+    func add(callback: @escaping BackendPurchaserInfoResponseHandler, key: String) -> CallbackCacheStatus {
+        return callbackQueue.sync { [self] in
+            var callbacksForKey = purchaserInfoCallbacksCache[key] ?? []
+            let requestAlreadyInFlight: CallbackCacheStatus = !callbacksForKey.isEmpty
```
maybe this `let` should now be called cacheStatus?

taquitos replied:
💯

aboedo reviewed (approved):
ready to go! I left a few comments that can be addressed later if you'd like

aboedo commented:
amazing work on this btw 💪
